File: Note/app/src/main/java/com/zukron/note/activity/DetailNoteActivity.java
package com.zukron.note.activity;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.widget.TextView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import com.jakewharton.threetenabp.AndroidThreeTen;
import com.orhanobut.dialogplus.DialogPlus;
import com.orhanobut.dialogplus.OnClickListener;
import com.orhanobut.dialogplus.ViewHolder;
import com.zukron.note.R;
import com.zukron.note.fragment.DefaultNoteFragment;
import com.zukron.note.fragment.ListNoteFragment;
import com.zukron.note.fragment.LongNoteFragment;
import com.zukron.note.model.DefaultNote;
import com.zukron.note.model.ListNote;
import com.zukron.note.model.Note;
import com.zukron.note.model.dao.DefaultNoteDao;
import com.zukron.note.model.dao.ListNoteDao;
import com.zukron.note.model.dao.NoteDao;
import com.zukron.note.util.Tools;

import org.threeten.bp.LocalDate;

public class DetailNoteActivity extends AppCompatActivity implements View.OnClickListener, OnClickListener {
    private MaterialToolbar mtCreate;
    private MaterialButton btnSaveCreate, btnDeleteCreate, btnFavCreate, btnMoreCreate;
    private TextView tvLastModifiedCreate, tvCreatedCreate;
    private CoordinatorLayout coordinatorLayout;
    private DialogPlus dialogPlus;
    private Window window;
    private DefaultNoteFragment defaultNoteFragment;
    private ListNoteFragment listNoteFragment;
    private NoteDao noteDao;
    private DefaultNoteDao defaultNoteDao;
    private ListNoteDao listNoteDao;
    private LongNoteFragment longNoteFragment;
    private int currentColor;
    public static final String EXTRA_NOTE = "extra_note";
    public static final String EXTRA_ACTION = "extra_action";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_note_detail);

        mtCreate = findViewById(R.id.mt_create);
        btnSaveCreate = findViewById(R.id.btn_save_create);
        btnDeleteCreate = findViewById(R.id.btn_delete_create);
        btnFavCreate = findViewById(R.id.btn_fav_create);
        btnMoreCreate = findViewById(R.id.btn_more_create);

        tvLastModifiedCreate = findViewById(R.id.tv_last_modified_create);
        tvCreatedCreate = findViewById(R.id.tv_created_create);

        coordinatorLayout = findViewById(R.id.coordinator_layout);

        // set dialog more menu
        setDialogMoreMenu();

        // set default note body
        setNoteBody();
    }

    private void setNoteBody() {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        // Note Data
        String action = getIntent().getStringExtra(DetailNoteActivity.EXTRA_ACTION);
        Note note = getIntent().getParcelableExtra(DetailNoteActivity.EXTRA_NOTE);

        if (action != null) {
            if (action.equals(Note.Action.insert)) { // insert
                defaultNoteFragment = new DefaultNoteFragment();
                fragmentTransaction.add(R.id.fl_note_create, defaultNoteFragment, DefaultNoteFragment.TAG);
            }

            if (action.equals(Note.Action.update)) { // update
                if (note != null) {
                    if (note.getType() == Note.Type.DefaultNote) {
                        defaultNoteFragment = DefaultNoteFragment.newInstance(note);
                        fragmentTransaction.add(R.id.fl_note_create, defaultNoteFragment, DefaultNoteFragment.TAG);
                    }

                    if (note.getType() == Note.Type.ListNote) {
                        listNoteFragment = ListNoteFragment.newInstance(note);
                        fragmentTransaction.add(R.id.fl_note_create, listNoteFragment, ListNoteFragment.TAG);
                    }

                    if (note.getType() == Note.Type.LongNote) {
                        longNoteFragment = LongNoteFragment.newInstance(note);
                        fragmentTransaction.add(R.id.fl_note_create, longNoteFragment, LongNoteFragment.TAG);
                    }
                }
            }
        }

        fragmentTransaction.commit();
    }

    private void changeNoteBody(String tag) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        if (tag.equals(DefaultNoteFragment.TAG)) {
            defaultNoteFragment = new DefaultNoteFragment();
            fragmentTransaction.replace(R.id.fl_note_create, defaultNoteFragment, DefaultNoteFragment.TAG);
        }

        if (tag.equals(ListNoteFragment.TAG)) {
            listNoteFragment = new ListNoteFragment();
            fragmentTransaction.replace(R.id.fl_note_create, listNoteFragment, ListNoteFragment.TAG);
        }

        if (tag.equals(LongNoteFragment.TAG)) {
            longNoteFragment = new LongNoteFragment();
            fragmentTransaction.replace(R.id.fl_note_create, longNoteFragment, LongNoteFragment.TAG);
        }

        fragmentTransaction.commit();
    }

    @Override
    protected void onStart() {
        super.onStart();

        mtCreate.setTitle("");
        setSupportActionBar(mtCreate);

        // window
        window = getWindow();

        // init background color
        initBackgroundColor();

        // set date
        setDateModifiedCreated();

        // button
        btnSaveCreate.setOnClickListener(this);
        btnDeleteCreate.setOnClickListener(this);
        btnFavCreate.setOnClickListener(this);
        btnMoreCreate.setOnClickListener(this);

        // database
        noteDao = new NoteDao(this);
        defaultNoteDao = new DefaultNoteDao(this);
        listNoteDao = new ListNoteDao(this);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        finish();
        return true;
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_save_create:
                insertUpdateAction();
                break;
            case R.id.btn_delete_create:
                deleteAction();
                break;
            case R.id.btn_fav_create:
                favAction();
                break;
            case R.id.btn_more_create:
                dialogPlus.show();
                break;
        }
    }

    @Override
    public void onClick(DialogPlus dialog, View view) {
        switch (view.getId()) {
            case R.id.btn_add_default_note:
                changeNoteBody(DefaultNoteFragment.TAG);
                break;
            case R.id.btn_add_list_note:
                changeNoteBody(ListNoteFragment.TAG);
                break;
            case R.id.btn_add_long_note:
                changeNoteBody(LongNoteFragment.TAG);
                break;
            case R.id.btn_pink_bg_create:
                setBackgroundColor(R.color.cardBgPink);
                break;
            case R.id.btn_blue_bg_create:
                setBackgroundColor(R.color.cardBgBlue);
                break;
            case R.id.btn_violet_bg_create:
                setBackgroundColor(R.color.cardBgViolet);
                break;
            case R.id.btn_orange_bg_create:
                setBackgroundColor(R.color.cardBgOrange);
                break;
        }

        dialog.dismiss();
    }

    private void initBackgroundColor() {
        Note note = getIntent().getParcelableExtra(DetailNoteActivity.EXTRA_NOTE);

        // change current color
        if (note == null) {
            currentColor = R.color.cardBgOrange;
        } else {
            currentColor = note.getColor();
        }

        setBackgroundColor(currentColor);
    }

    private void setDateModifiedCreated() {
        String action = getIntent().getStringExtra(DetailNoteActivity.EXTRA_ACTION);
        Note note = getIntent().getParcelableExtra(DetailNoteActivity.EXTRA_NOTE);

        if (action != null) {
            if (action.equals(Note.Action.insert)) {
                tvLastModifiedCreate.setText("");
                tvCreatedCreate.setText("");
            } else if (action.equals(Note.Action.update)) {
                if (note != null) {
                    tvLastModifiedCreate.setText(Tools.formatDate(note.getModified()));
                    tvCreatedCreate.setText(Tools.formatDate(note.getCreated()));
                }
            }
        }
    }

    private void setDialogMoreMenu() {
        dialogPlus = DialogPlus.newDialog(this)
                .setExpanded(true, 600)
                .setGravity(Gravity.TOP)
                .setExpanded(true)
                .setOnClickListener(this)
                .setContentHolder(new ViewHolder(R.layout.more_menu))
                .create();
    }

    private void setBackgroundColor(int color) {
        currentColor = color;
        coordinatorLayout.setBackgroundColor(ContextCompat.getColor(this, color));

        // status bar and button save
        window.setStatusBarColor(ContextCompat.getColor(this, color));
        btnSaveCreate.setTextColor(ContextCompat.getColor(this, color));
    }

    private void favAction() {
        btnFavCreate.setChecked(btnFavCreate.isChecked());

        if (btnFavCreate.isChecked()) {
            btnFavCreate.setIcon(getDrawable(R.drawable.ic_baseline_favorite_border_24));
        } else {
            btnFavCreate.setIcon(getDrawable(R.drawable.ic_baseline_favorite_24));
        }
    }

    private void deleteAction() {
        final Note note = getIntent().getParcelableExtra(DetailNoteActivity.EXTRA_NOTE);

        if (note != null) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Hapus");
            builder.setMessage("Apakah yakin hapus ?");
            builder.setPositiveButton("Iya", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    int id = note.getId();
                    noteDao.delete(id);

                    if (note.getType() == Note.Type.DefaultNote)
                        defaultNoteDao.delete(id);

                    if (note.getType() == Note.Type.ListNote)
                        listNoteDao.delete(id);

                    finish();
                }
            });
            builder.setNegativeButton("Tidak", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    dialogInterface.dismiss();
                }
            });

            builder.show();
        }
    }

    private void insertUpdateAction() {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.fl_note_create);
        String action = getIntent().getStringExtra(DetailNoteActivity.EXTRA_ACTION);
        AndroidThreeTen.init(this);

        if (fragment instanceof DefaultNoteFragment) {
            DefaultNote defaultNote = defaultNoteFragment.getNote();

            if (defaultNote != null && action != null) {
                if (action.equals(Note.Action.insert)) {
                    Note note = new Note(defaultNote.getId(), defaultNote.getTitle(), Note.Type.DefaultNote, LocalDate.now(), LocalDate.now(), currentColor);

                    noteDao.insert(note);
                    defaultNoteDao.insert(defaultNote);
                }

                if (action.equals(Note.Action.update)) {
                    Note note = new Note(defaultNote.getId(), defaultNote.getTitle(), LocalDate.now(), currentColor);

                    noteDao.update(note);
                    defaultNoteDao.update(defaultNote);
                }

                finish();
            }
        }

        if (fragment instanceof ListNoteFragment) {
            ListNote listNote = listNoteFragment.getNote();

            if (listNote != null && action != null) {
                if (action.equals(Note.Action.insert)) {
                    Note note = new Note(listNote.getId(), listNote.getTitle(), Note.Type.ListNote, LocalDate.now(), LocalDate.now(), currentColor);

                    noteDao.insert(note);
                    listNoteDao.insert(listNote);
                }

                if (action.equals(Note.Action.update)) {
                    Note note = new Note(listNote.getId(), listNote.getTitle(), LocalDate.now(), currentColor);

                    noteDao.update(note);
                    listNoteDao.update(listNote);
                }

                finish();
            }
        }

        if (fragment instanceof LongNoteFragment) {
            DefaultNote defaultNote = longNoteFragment.getNote();

            if (defaultNote != null && action != null) {
                if (action.equals(Note.Action.insert)) {
                    Note note = new Note(defaultNote.getId(), defaultNote.getTitle(), Note.Type.LongNote, LocalDate.now(), LocalDate.now(), currentColor);

                    noteDao.insert(note);
                    defaultNoteDao.insert(defaultNote);
                }

                if (action.equals(Note.Action.update)) {
                    Note note = new Note(defaultNote.getId(), defaultNote.getTitle(), LocalDate.now(), currentColor);

                    noteDao.update(note);
                    defaultNoteDao.update(defaultNote);
                }

                finish();
            }
        }
    }
}

File: Note/app/src/main/java/com/zukron/note/activity/DetailNoteActivity.kt
package com.zukron.note.activity

import android.os.Bundle
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.jakewharton.threetenabp.AndroidThreeTen
import com.orhanobut.dialogplus.DialogPlus
import com.orhanobut.dialogplus.OnClickListener
import com.orhanobut.dialogplus.ViewHolder
import com.zukron.note.R
import com.zukron.note.fragment.DefaultNoteFragment
import com.zukron.note.fragment.ListNoteFragment
import com.zukron.note.fragment.LongNoteFragment
import com.zukron.note.model.Note
import com.zukron.note.model.dao.DefaultNoteDaoJava
import com.zukron.note.model.dao.ListNoteDaoJava
import com.zukron.note.model.dao.NoteDaoJava
import com.zukron.note.util.Utilities.formatToString
import kotlinx.android.synthetic.main.activity_note_detail.*
import org.threeten.bp.LocalDate

/**
 * Project name is Note
 * Created by Zukron Alviandy R on 9/4/2020
 * Contact me if any issues on zukronalviandy@gmail.com
 */
class DetailNoteActivity : AppCompatActivity(), OnClickListener, View.OnClickListener {
    companion object {
        const val EXTRA_NOTE = "extra_note"
        const val EXTRA_ACTION = "extra_action"
    }

    private lateinit var dialogPlus: DialogPlus
    private lateinit var defaultNoteFragment: DefaultNoteFragment
    private lateinit var listNoteFragment: ListNoteFragment
    private lateinit var longNoteFragment: LongNoteFragment
    private lateinit var noteDaoJava: NoteDaoJava
    private lateinit var defaultNoteDaoJava: DefaultNoteDaoJava
    private lateinit var listNoteDaoJava: ListNoteDaoJava
    private var currentColor: Int = R.color.cardBgOrange
    private val setBackgroundColor: (color: Int) -> Unit = {
        currentColor = it
        coordinator_layout.setBackgroundColor(
                ContextCompat.getColor(this, it)
        )

        // status bar and button save color
        val window = window
        window.statusBarColor = ContextCompat.getColor(
                this, it
        )
        btn_save.setTextColor(ContextCompat.getColor(
                this, it
        ))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_note_detail)

        // build dialog plus
        dialogPlus = DialogPlus.newDialog(this)
                .setExpanded(true)
                .setGravity(Gravity.TOP)
                .setOnClickListener(this)
                .setContentHolder(ViewHolder(R.layout.more_menu))
                .create()

        // toolbar
        material_toolbar.title = ""
        setSupportActionBar(material_toolbar)

        // change background color and set text date modified
        val action = intent.getStringExtra(EXTRA_ACTION)
        val note = intent.getParcelableExtra<Note>(EXTRA_NOTE)
        note.let {
            if (action != null) {
                if (action == Note.Action.insert) {
                    tv_last_modified.text = ""
                    tv_created.text = ""
                } else if (action == Note.Action.update) {
                    if (it != null) {
                        // change background
                        currentColor = it.color

                        // text view
                        tv_last_modified.text = it.modifiedDate.formatToString()
                        tv_created.text = it.createdDate.formatToString()
                    }
                }
            }
        }
        setBackgroundColor(currentColor)

        // set note body
        setNoteBody()

        // init database
        noteDaoJava = NoteDaoJava(this)
        defaultNoteDaoJava = DefaultNoteDaoJava(this)
        listNoteDaoJava = ListNoteDaoJava(this)

        // button listener
        btn_save.setOnClickListener(this)
        btn_delete.setOnClickListener(this)
        btn_fav.setOnClickListener(this)
        btn_more.setOnClickListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        finish()
        return true
    }

    override fun onClick(dialog: DialogPlus?, view: View?) {
        when (view?.id) {
            R.id.btn_add_default_note -> changeNoteBody(DefaultNoteFragment.TAG)
            R.id.btn_add_list_note -> changeNoteBody(ListNoteFragment.TAG)
            R.id.btn_add_long_note -> changeNoteBody(LongNoteFragment.TAG)
            R.id.btn_pink_bg -> setBackgroundColor(R.color.cardBgPink)
            R.id.btn_blue_bg -> setBackgroundColor(R.color.cardBgBlue)
            R.id.btn_violet_bg -> setBackgroundColor(R.color.cardBgViolet)
            R.id.btn_orange_bg -> setBackgroundColor(R.color.cardBgOrange)
        }

        dialog?.dismiss()
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            R.id.btn_save -> insertUpdateAction()
            R.id.btn_delete -> deleteAction()
            R.id.btn_fav -> favAction()
            R.id.btn_more -> dialogPlus.show()
        }
    }

    private fun setNoteBody() {
        val fragmentTransaction = supportFragmentManager.beginTransaction()

        // note data
        val action = intent.getStringExtra(EXTRA_ACTION)
        val note = intent.getParcelableExtra<Note>(EXTRA_NOTE)

        if (action != null) {
            // action = insert
            if (action == Note.Action.insert) {
                defaultNoteFragment = DefaultNoteFragment()
                fragmentTransaction.add(R.id.fl_note, defaultNoteFragment, DefaultNoteFragment.TAG)
            }

            // action = update
            if (action == Note.Action.update) {
                if (note != null) {
                    if (note.type == Note.Type.defaultNote) {
                        defaultNoteFragment = DefaultNoteFragment.newInstance(note)
                        fragmentTransaction.add(R.id.fl_note, defaultNoteFragment, DefaultNoteFragment.TAG)
                    }

                    if (note.type == Note.Type.listNote) {
                        listNoteFragment = ListNoteFragment.newInstance(note)
                        fragmentTransaction.add(R.id.fl_note, listNoteFragment, ListNoteFragment.TAG)
                    }

                    if (note.type == Note.Type.longNote) {
                        longNoteFragment = LongNoteFragment.newInstance(note)
                        fragmentTransaction.add(R.id.fl_note, longNoteFragment, LongNoteFragment.TAG)
                    }
                }
            }

            fragmentTransaction.commit()
        }
    }

    private fun changeNoteBody(tag: String) {
        val fragmentTransaction = supportFragmentManager.beginTransaction()

        if (tag == DefaultNoteFragment.TAG) {
            defaultNoteFragment = DefaultNoteFragment()
            fragmentTransaction.replace(R.id.fl_note, defaultNoteFragment, DefaultNoteFragment.TAG)
        }

        if (tag == ListNoteFragment.TAG) {
            listNoteFragment = ListNoteFragment()
            fragmentTransaction.replace(R.id.fl_note, listNoteFragment, ListNoteFragment.TAG)
        }

        if (tag == LongNoteFragment.TAG) {
            longNoteFragment = LongNoteFragment()
            fragmentTransaction.replace(R.id.fl_note, longNoteFragment, LongNoteFragment.TAG)
        }

        fragmentTransaction.commit()
    }

    private fun insertUpdateAction() {
        val fragment = supportFragmentManager.findFragmentById(R.id.fl_note)
        val action = intent.getStringExtra(EXTRA_ACTION)
        AndroidThreeTen.init(this)

        if (fragment is DefaultNoteFragment) {
            val defaultNote = defaultNoteFragment.getNote()

            if (defaultNote != null && action != null) {
                val note = Note(
                        id = defaultNote.id,
                        title = defaultNote.title,
                        type = Note.Type.defaultNote,
                        modifiedDate = LocalDate.now(),
                        createdDate = LocalDate.now(),
                        color = currentColor
                )

                if (action == Note.Action.insert) {
                    noteDaoJava.insert(note)
                    defaultNoteDaoJava.insert(defaultNote)
                }

                if (action == Note.Action.update) {
                    noteDaoJava.update(note)
                    defaultNoteDaoJava.update(defaultNote)
                }

                finish()
            }
        }

        if (fragment is ListNoteFragment) {
            val listNote = listNoteFragment.getNote()

            if (listNote != null && action != null) {
                val note = Note(
                        id = listNote.id,
                        title = listNote.title,
                        type = Note.Type.listNote,
                        modifiedDate = LocalDate.now(),
                        createdDate = LocalDate.now(),
                        color = currentColor
                )

                if (action == Note.Action.insert) {
                    noteDaoJava.insert(note)
                    listNoteDaoJava.insert(listNote)
                }

                if (action == Note.Action.update) {
                    noteDaoJava.update(note)
                    listNoteDaoJava.update(listNote)
                }

                finish()
            }
        }

        if (fragment is LongNoteFragment) {
            val defaultNote = longNoteFragment.getNote()

            if (defaultNote != null && action != null) {
                val note = Note(
                        id = defaultNote.id,
                        title = defaultNote.title,
                        type = Note.Type.longNote,
                        modifiedDate = LocalDate.now(),
                        createdDate = LocalDate.now(),
                        color = currentColor
                )

                if (action == Note.Action.insert) {
                    noteDaoJava.insert(note)
                    defaultNoteDaoJava.insert(defaultNote)
                }

                if (action == Note.Action.update) {
                    noteDaoJava.update(note)
                    defaultNoteDaoJava.update(defaultNote)
                }

                finish()
            }
        }
    }

    private fun deleteAction() {
        val note = intent.getParcelableExtra<Note>(EXTRA_NOTE)

        if (note != null) {
            val alertDialog = AlertDialog.Builder(this)
            alertDialog.setTitle("Delete")
            alertDialog.setMessage("Are you sure ?")
            alertDialog.setPositiveButton("Yes") { _, _ ->
                val id = note.id
                noteDaoJava.delete(id)

                if (note.type == Note.Type.defaultNote) defaultNoteDaoJava.delete(id)
                if (note.type == Note.Type.listNote) listNoteDaoJava.delete(id)

                finish()
            }

            alertDialog.setNegativeButton("No") { dialogInterface, _ ->
                dialogInterface.dismiss()
            }

            alertDialog.show()
        }
    }

    private fun favAction() {
        btn_fav.isChecked = btn_fav.isChecked

        val icon: Int = if (btn_fav.isChecked) {
            R.drawable.ic_baseline_favorite_border_24
        } else {
            R.drawable.ic_baseline_favorite_24
        }

        btn_fav.icon = ContextCompat.getDrawable(
                this, icon
        )
    }
}